strawberryfields/ops.py

```diff
@@ -1808,6 +1808,98 @@ def _decompose(self, reg, **kwargs):
+
+    where :math:`B` is a :math:N/2\times N/2` matrix representing the (weighted)
+    edges between the vertex set.
+
+    This operation uses the Takagi decomposition to decompose
+    an adjacency matrix into a sequence of two mode squeezers, beamsplitters, and
+    rotation gates.
+
+    Args:
+        A (array): Either an :math:`N\times N` complex or real symmetric adjacency matrix
+            :math:`A`, or an :math:`N/2\times N/2` complex or real matrix :math:`B`
+            representing the edges between the vertex sets if ``edges=True``.
+        mean_photon (float): guarantees that the mean photon number in the pure Gaussian state
+            representing the graph satisfies  :math:`\frac{1}{N}\sum_{i=1}^N sinh(r_{i})^2 ==` :code:``mean_photon``
+        edges (bool): set to ``True`` if argument ``A`` represents the edges :math:`B`
+            between the vertex sets rather than the full adjacency matrix
+        drop_identity (bool): If ``True``, decomposed gates with trivial parameters,
+            such that they correspond to an identity operation, are removed.
+        tol (float): the tolerance used when checking if the input matrix is symmetric:
+            :math:`|A-A^T| <` tol
+    """
+
+    def __init__(self, A, mean_photon=1.0, edges=False, drop_identity=True, tol=1e-6):
+        self.mean_photon = mean_photon
+        self.tol = tol
+        self.identity = np.all(np.abs(A - np.identity(len(A))) < _decomposition_merge_tol)
+        self.drop_identity = drop_identity
+
+        if edges:
+            self.ns = 2*A.shape[0]
+            B = A
+        else:
+            self.ns = A.shape[0]
+
+            # check if A is a bipartite graph
+            N = A.shape[0]//2
+            A00 = A[:N, :N]
+            A11 = A[N:, N:]
+
+            diag_zeros = np.allclose(A00, np.zeros_like(A00), atol=tol, rtol=0)
+            diag_zeros = diag_zeros and np.allclose(A11, np.zeros_like(A11), atol=tol, rtol=0)
+
+            if (not diag_zeros) or (not np.allclose(A, A.T, atol=tol, rtol=0)):
+                raise ValueError("Adjacency matrix {} does not represent a bipartite graph".format(A))
+
+            B = A[:N, N:]
+
+        super().__init__([B])
+
+    def _decompose(self, reg, **kwargs):
+        mean_photon = kwargs.get("mean_photon", self.mean_photon)
+        tol = kwargs.get("tol", self.tol)
+        mesh = kwargs.get("mesh", "rectangular")
+        drop_identity = kwargs.get("drop_identity", self.drop_identity)
+
+        cmds = []
+
+        B = self.p[0].x
+        N = len(B)
+
+        sq, U, V = dec.bipartite_graph_embed(B, mean_photon_per_mode=mean_photon, atol=tol, rtol=0)
+
+        if not self.identity or not drop_identity:
+            for m, s in enumerate(sq):
+                s = s if np.abs(s) >= _decomposition_tol else 0
+
+                if not (drop_identity and s == 0):
+                    cmds.append(Command(S2gate(-s), (reg[m], reg[m+N])))
+
```

Comment:
The below could possibly be refactored to reduce code repetition, e.g.:
```python
for X, _reg in ((U, reg[:N]), (V, reg[N:]):
    if np.allclose(X, np.identity(len(X)), atol=_decomposition_tol, rtol=0):
        X = np.identity(len(X))
    if not (drop_identity and np.all(X == np.identity(len(X)))):
        cmds.append(Command(Interferometer(X, mesh=mesh, drop_identity=drop_identity, tol=tol), _reg))
```

Reply:
Thanks! Made this change.